File: scripts/05_generate_longitudinal_mock.py
import csv
import random
import os

# Set seed for reproducibility
random.seed(42)

n_subjects = 10
n_timepoints = 10
features = ["Feature_A", "Feature_B", "Feature_C"]

data = []

for feature in features:
    for subject in range(1, n_subjects + 1):
        subj_id = f"S{subject:02d}"
        
        # Determine dynamics for this subject/feature
        if feature == "Feature_A":
            # Transition from low (~0.2) to high (~0.8) at T5
            # Except for S08 who stays low
            if subj_id == "S08":
                base_values = [random.uniform(0.1, 0.3) for _ in range(n_timepoints)]
            else:
                base_values = [random.uniform(0.1, 0.3) for _ in range(4)] + \
                              [random.uniform(0.7, 0.9) for _ in range(6)]
        elif feature == "Feature_B":
            # Stable state around 0.5
            base_values = [random.uniform(0.45, 0.55) for _ in range(n_timepoints)]
        else:
            # Random noise 0-1
            base_values = [random.uniform(0.0, 1.0) for _ in range(n_timepoints)]
            
        for t in range(1, n_timepoints + 1):
            val = base_values[t-1]
            
            # Add missing values for S05 at T3 and T7
            if subj_id == "S05" and t in [3, 7]:
                val = "" # Representing NA in CSV
            
            # Add random drops (approx 3% of data)
            elif random.random() < 0.03:
                val = ""
                
            data.append({
                "Timepoint": t,
                "Subject_ID": subj_id,
                "Feature": feature,
                "Value": val
            })

# Ensure directory exists
os.makedirs("data", exist_ok=True)

# Save to CSV using csv module
output_path = "data/mock_longitudinal_transitions.csv"
with open(output_path, 'w', newline='') as csvfile:
    fieldnames = ["Timepoint", "Subject_ID", "Feature", "Value"]
    writer = csv.DictWriter(csvfile, fieldnames=fieldnames)

    writer.writeheader()
    for row in data:
        writer.writerow(row)

print(f"Mock longitudinal data generated: {output_path}")
print(f"Schema: {fieldnames}")
print(f"Scale: {n_subjects} subjects, {n_timepoints} timepoints, {len(features)} features.")

File: scripts/05_generate_longitudinal_mock.R
#!/usr/bin/env Rscript
# Description: Generates longitudinal mock data for transition dynamics analysis.
# Schema: Timepoint, Subject_ID, Feature, Value

set.seed(42)
library(dplyr)
library(tidyr)

n_subjects <- 10
n_timepoints <- 10
features <- c("Feature_A", "Feature_B", "Feature_C")

# Create a grid of all combinations
df <- expand.grid(
  Subject_ID = sprintf("S%02d", 1:n_subjects),
  Timepoint = 1:n_timepoints,
  Feature = features
)

# Generate values based on dynamics
df <- df %>%
  mutate(
    Value = case_when(
      # Feature_A: Shifting from low to high at T5 (mostly)
      Feature == "Feature_A" ~ ifelse(Timepoint < 5, 
                                      runif(n(), 0.1, 0.3), 
                                      runif(n(), 0.7, 0.9)),
      
      # Feature_B: Stable state
      Feature == "Feature_B" ~ runif(n(), 0.45, 0.55),
      
      # Feature_C: Random noise
      Feature == "Feature_C" ~ runif(n(), 0.0, 1.0)
    )
  )

# Add Edge Cases
# 1. Subject S08 is stable for Feature_A (no transition)
df$Value[df$Subject_ID == "S08" & df$Feature == "Feature_A"] <- runif(sum(df$Subject_ID == "S08" & df$Feature == "Feature_A"), 0.1, 0.3)

# 2. Add Missing Values (NAs)
# S05 is missing some T3 and T7
df$Value[df$Subject_ID == "S05" & df$Timepoint %in% c(3, 7)] <- NA

# 3. Add some random drops for robustness testing
drop_indices <- sample(1:nrow(df), 10)
df$Value[drop_indices] <- NA

# Save to CSV
dir.create("data", showWarnings = FALSE)
write.csv(df, "data/mock_longitudinal_transitions.csv", row.names = FALSE)

message("Mock longitudinal data generated: data/mock_longitudinal_transitions.csv")
message("Schema: ", paste(colnames(df), collapse = ", "))
message("Scale: ", n_subjects, " subjects, ", n_timepoints, " timepoints, ", length(features), " features.")
